drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,443 @@
+};
+
+
+/******************************************************
+ *                 Function
+ ******************************************************/
+
+int airoc_wifi_power_on(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, wifi_reg_on_gpios)
+	int ret;
+	const struct airoc_wifi_config *config = dev->config;
+
+	/* Check WIFI REG_ON gpio instance */
+	if (!device_is_ready(config->wifi_reg_on_gpio.port)) {
+		LOG_ERR("Error: failed to configure wifi_reg_on %s pin %d",
+			config->wifi_reg_on_gpio.port->name, config->wifi_reg_on_gpio.pin);
+		return -EIO;
+	}
+
+	/* Configure wifi_reg_on as output  */
+	ret = gpio_pin_configure_dt(&config->wifi_reg_on_gpio, GPIO_OUTPUT);
+	if (ret) {
+		LOG_ERR("Error %d: failed to configure wifi_reg_on %s pin %d", ret,
+			config->wifi_reg_on_gpio.port->name, config->wifi_reg_on_gpio.pin);
+		return ret;
+	}
+	ret = gpio_pin_set_dt(&config->wifi_reg_on_gpio, 0);
+	if (ret) {
+		return ret;
+	}
+
+	/* Allow CBUCK regulator to discharge */
+	(void)cyhal_system_delay_ms(WLAN_CBUCK_DISCHARGE_MS);
+
+	/* WIFI power on */
+	ret = gpio_pin_set_dt(&config->wifi_reg_on_gpio, 1);
+	if (ret) {
+		return ret;
+	}
+	(void)cyhal_system_delay_ms(WLAN_POWER_UP_DELAY_MS);
+#endif /* DT_INST_NODE_HAS_PROP(0, reg_on_gpios) */
+
+	return 0;
+}
+
+int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface,
+			    whd_netif_funcs_t *netif_funcs, whd_buffer_funcs_t *buffer_if)
+{
+	int ret;
+	struct airoc_wifi_data *data = dev->data;
+	const struct airoc_wifi_config *config = dev->config;
+
+	whd_sdio_config_t whd_sdio_config = {
+		.sdio_1bit_mode = WHD_FALSE,
+		.high_speed_sdio_clock = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
+	whd_oob_config_t oob_config = {
+		.host_oob_pin = (void *)&config->wifi_host_wake_gpio,
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_sdio_config.oob_config = oob_config;
+#endif
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on retuens fail");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->sdhc_dev)) {
+		LOG_ERR("SDHC device is not ready");
+		return -ENODEV;
+	}
+
+	ret = sd_init(config->sdhc_dev, &data->card);
+	if (ret) {
+		return ret;
+	}
+
+	cy_rslt_t whd_ret = whd_init(&data->whd_drv, &init_config_default, &resource_ops, buffer_if,
+				     netif_funcs);
+	if (whd_ret == CY_RSLT_SUCCESS) {
+		whd_ret = whd_bus_sdio_attach(data->whd_drv, &whd_sdio_config,
+					      (whd_sdio_t)&data->card);
+
+		if (whd_ret == CY_RSLT_SUCCESS) {
+			whd_ret = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (whd_ret != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+			return -ENODEV;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Implement SDIO CMD52/53 wrappers
+ */
+whd_result_t whd_bus_sdio_cmd52(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
```

Comment:
Is there a reason you are interfacing directly with the SDHC driver layer here via `sdhc_request`? The expectation is that SDIO drivers would implement CMD52 by calling `sdio_read_byte` or `sdio_write_byte` depending on the command direction.

Reply:
Looking at this more- I see that in the SDIO PR you mentioned your plan to do this:

> Implementation of SDIO CMD53/52 (will use SDHC sdhc_request function. The core functions sdio_io_rw_direct/sdio_io_rw_extended from SD/SDIO subsystem are static, so i have to SDHC sdhc_request function for WHD implementation).

while `sdio_io_rw_direct` and `sdio_io_rw_extended` are static, the `sdio_read_byte`/`sdio_write_byte` functions provide wrappers around those functions that should be usable by an external driver. Is there something specific that prevents you from doing so?